Sources/FoundationEssentials/Data/Data+Reading.swift

```diff
@@ -167,12 +168,126 @@ struct ReadBytesResult {
     var deallocator: Data.Deallocator?
 }
 
+#if os(Windows)
+private func read(from hFile: HANDLE, at path: PathOrURL,
+                  into pBuffer: UnsafeMutableRawPointer, length dwLength: Int,
+                  chunkSize dwChunk: Int = 4096, progress bProgress: Bool)
+        throws -> Int {
+    var pBuffer = pBuffer
+    let progress = bProgress && Progress.current() != nil ? Progress(totalUnitCount: Int64(dwLength)) : nil
+
+    var dwBytesRemaining: DWORD = DWORD(dwLength)
+    while dwBytesRemaining > 0 {
+        if let progress, progress.isCancelled {
+            throw CocoaError(.userCancelled)
+        }
+
+        let dwBytesToRead: DWORD =
+            DWORD(clamping: DWORD(min(DWORD(dwChunk), dwBytesRemaining)))
+        var dwBytesRead: DWORD = 0
+        if !ReadFile(hFile, pBuffer, dwBytesToRead, &dwBytesRead, nil) {
+            throw CocoaError.errorWithFilePath(path, win32: GetLastError(), reading: true)
+        }
+        dwBytesRemaining -= DWORD(clamping: dwBytesRead)
+        progress?.completedUnitCount = Int64(dwLength - Int(dwBytesRemaining))
+        if dwBytesRead < dwBytesToRead {
+            break
+        }
+        pBuffer = pBuffer.advanced(by: Int(dwBytesRead))
+    }
+    return dwLength - Int(dwBytesRemaining)
+}
+#endif
+
 internal func readBytesFromFile(path inPath: PathOrURL, reportProgress: Bool, maxLength: Int?, options: Data.ReadingOptions, attributesToRead: [String], attributes: inout [String: Data]) throws -> ReadBytesResult {
     if inPath.isEmpty {
         // For compatibility, throw a different error than the perhaps-expected 'file not found' here (41646641)
         throw CocoaError(.fileReadInvalidFileName)
     }
-    
+
+#if os(Windows)
+    let hFile: HANDLE = inPath.withFileSystemRepresentation {
+        String(cString: $0!).withCString(encodedAs: UTF16.self) {
```

Comment:
Given that Windows uses UTF-16 for the file system representation, I think we should probably just make `withFileSystemRepresentation` provide a UTF-16 pointer on Windows so that callers don't have to do this every time. Thoughts? The linux code path in `String+Path.swift` already just basically equates to a call to `withCString` so I think we could just make a new path for Windows that calls `withCString(encodedAs: UTF16.self)` and then just ensure the function signatures use the right types

Reply:
Ah ... so, we now run into the critical juncture of - how do we make code portable? If the signature is different between windows and !windows - the body will need to be replaced. For internal uses, I think that I want to never use `withFileSystemRepresentation` in Windows specific paths and prefer to use `withNTPathRepresentation`. Let me upload that as a splintered change so that we can start using it and setup context.

Reply:
Put up #531 so that we can discuss the windows path representation first.

Reply:
Okay, we can now convert a `String` type to the proper spelling for a NT path, however, we still do not handle `URL` properly.